keps/sig-network/1669-proxy-terminating-endpoints/README.md

```diff
@@ -0,0 +1,377 @@
+
+When the traffic policy is "Cluster" and some endpoints are terminating, all traffic should be routed to the ready endpoints that are not terminating..
+
+### Example: only some endpoints terminating on a node when traffic policy is "Local"
+
+When the traffic policy is "Local" and some endpoints are terminating within a single node, traffic should be routed to ready endpoints on that node that are not terminating.
+
+### Example: all endpoints terminating and traffic policy is "Cluster"
+
+When the traffic policy is "Cluster" and all endpoints are terminating, then traffic should be routed to any terminating endpoint that is ready.
+
+### Example: all endpoints terminating on a node when traffic policy is "Local"
+
+When the traffic policy is "Local" and all endpoints are terminating within a single node, then traffic should be routed to any terminating endpoint that is ready on that node.
+
+
+### Handling terminating endpoints that are not ready.
+
+It is worth noting that traffic should not be routed to terminating pods if their readiness probe is failing, even if it is the only endpoints remaining. This is to give workloads
+the flexibility/control to opt out of this behavior by either exiting immediately or failing the readiness probe when receiving SIGTERM from kubelet. This would also be counter-intuitive
+to the current understanding of readiness probes.
+
+
+### User Stories (optional)
+
+#### Story 1
+
+As a user I would like to do a rolling update of a Deployment fronted by a Service Type=LoadBalancer with ExternalTrafficPolicy=Local.
+If a node that has only 1 pod of said deployment goes into the `Terminating` state, all traffic to that node is dropped until either a new pod
+comes up or my cloud provider removes the node from the loadbalancer's node pool. Ideally the terminating pod should gracefully handle traffic to this node
+until either one of the conditions are satisfied.
+
+### Risks and Mitigations
+
+There are scalability implications to tracking termination state in EndpointSlice. For now we are assuming that the performance trade-offs are worthwhile but
+future testing may change this decision. See KEP 1672 for more details.
+
+## Design Details
+
+### Additions to EndpointSlice
+
+This work depends on the `Terminating` condition existing on the EndpointSlice API (see KEP 1672) in order to check the termination state of an endpoint.
+
+### kube-proxy
+
+Updates to kube-proxy when watching EndpointSlice:
+* update kube-proxy endpoints info to track terminating endpoints based on endpoint.condition.terminating in EndpointSlice.
+* update kube-proxy endpoints info to track endpoint readiness based on endpoint.condition.ready in EndpointSlice
+* within the scope of the traffic policy for a Service, iterate the following set of endpoints, picking the first set that has at least 1 ready endpoint:
+  * ready endpoints that are not terminating
+  * ready endpoints that are terminating
```

Comment:
Worth noting here that in the previous version of this KEP, we said we would fall back to not-ready terminating endpoints as a last resort. After thinking about this more, I think it's better to only fall back to ready terminating endpoints. It's generally more predictable and safer IMO. Thoughts?

Reply:
SGTM